numeric.c

```diff
@@ -771,36 +771,29 @@ Perl_grok_infnan(pTHX_ const char** sp, const char* send)
         s++; if (s == send || isALPHA_FOLD_NE(*s, 'N')) return 0;
         s++; if (s == send) return 0;
         if (isALPHA_FOLD_EQ(*s, 'F')) {
-            s++;
+            flags |= IS_NUMBER_INFINITY | IS_NUMBER_NOT_INT;
+            *sp = ++s;
             if (s < send && (isALPHA_FOLD_EQ(*s, 'I'))) {
-                int fail =
-                    flags | IS_NUMBER_INFINITY | IS_NUMBER_NOT_INT | IS_NUMBER_TRAILING;
-                s++; if (s == send || isALPHA_FOLD_NE(*s, 'N')) return fail;
-                s++; if (s == send || isALPHA_FOLD_NE(*s, 'I')) return fail;
-                s++; if (s == send || isALPHA_FOLD_NE(*s, 'T')) return fail;
-                s++; if (s == send || isALPHA_FOLD_NE(*s, 'Y')) return fail;
-                s++;
+                int trail = flags | IS_NUMBER_TRAILING;
+                s++; if (s == send || isALPHA_FOLD_NE(*s, 'N')) return trail;
+                s++; if (s == send || isALPHA_FOLD_NE(*s, 'I')) return trail;
+                s++; if (s == send || isALPHA_FOLD_NE(*s, 'T')) return trail;
+                s++; if (s == send || isALPHA_FOLD_NE(*s, 'Y')) return trail;
+                *sp = ++s;
             } else if (odh) {
                 while (s < send && *s == '0') { /* 1.#INF00 */
                     s++;
                 }
             }
-            while (s < send && isSPACE(*s))
-                s++;
-            if (s < send && *s) {
-                flags |= IS_NUMBER_TRAILING;
-            }
-            flags |= IS_NUMBER_INFINITY | IS_NUMBER_NOT_INT;
+            goto ok_check_space;
         }
         else if (isALPHA_FOLD_EQ(*s, 'D') && odh) { /* 1.#IND */
             s++;
             flags |= IS_NUMBER_NAN | IS_NUMBER_NOT_INT;
             while (s < send && *s == '0') { /* 1.#IND00 */
                 s++;
             }
-            if (s < send && *s) {
-                flags |= IS_NUMBER_TRAILING;
-            }
+            goto ok_check_space;
```

Comment:
Are these `goto ok_check_space;`'s really needed?  Enclosing blocks seem to naturally exit to `ok_check_space:` even without them.

Reply:
I think they're valuable for consistency: the intent is that `goto ok_check_space` confirms we have met its prerequisites: "We parsed something valid, s points after it, flags describes it", hence also the `/* NOTREACHED */` to discourage simply falling out of the block. I'd expect any C compiler to generate the same code either way, this is just for the benefit of humans.

Reply:
Oh, that makes sense.
Then, how about using `NOT_REACHED` macro in addition to `/* NOTREACHED */`?

Reply:
> Oh, that makes sense. Then, how about using `NOT_REACHED` macro in addition to `/* NOTREACHED */`?

Ah yes, I had forgotten that we also had a macro for it. Now added, and also updated hv/gh19010-rebase.